go/merkletree2/tree_test.go

```diff
@@ -746,6 +752,317 @@ func TestNodeEncodingBasic(t *testing.T) {
+	startSeqno := Seqno(rand.Intn(int(maxSeqno)-1) + 1)
+	endSeqno := startSeqno
+
+	eProof, err := tree.GetExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, endSeqno)
+	require.NoError(t, err)
+
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &eProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.NoError(t, err)
+
+	k := keys[rand.Intn(len(keys))]
+
+	kvp, ieProof, err := tree.GetKeyValuePairWithInclusionExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, endSeqno, k)
+	require.NoError(t, err)
+	require.Equal(t, k, kvp.Key)
+
+	err = verifier.VerifyInclusionExtensionProof(NewLoggerContextTodoForTesting(t), kvp, &ieProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.NoError(t, err)
+
+	endSeqno = startSeqno + 1
+
+	eProof, err = tree.GetExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, endSeqno)
+	require.NoError(t, err)
+
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &eProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.NoError(t, err)
+
+	kvp, ieProof, err = tree.GetKeyValuePairWithInclusionExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, endSeqno, k)
+	require.NoError(t, err)
+	require.Equal(t, k, kvp.Key)
+
+	err = verifier.VerifyInclusionExtensionProof(NewLoggerContextTodoForTesting(t), kvp, &ieProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.NoError(t, err)
+}
+
+func TestExtensionProofsFailureBranches(t *testing.T) {
+	cfg, err := NewConfig(SHA512_256Encoder{}, false, 1, 1, 3, ConstructStringValueContainer)
+	require.NoError(t, err)
+
+	// make test deterministic
+	rand.Seed(1)
+
+	tree, err := NewTree(cfg, 2, NewInMemoryStorageEngine(cfg), RootVersionTesting)
+	require.NoError(t, err)
+
+	keys, err := makeRandomKeysForTesting(uint(cfg.KeysByteLength), 5)
+	require.NoError(t, err)
+
+	rootHashes := make(map[Seqno]Hash)
+
+	maxSeqno := Seqno(100)
+	// build a bunch of tree versions:
+	for j := Seqno(1); j < maxSeqno; j++ {
+		kvps, err := makeRandomKVPFromKeysForTesting(keys)
+		require.NoError(t, err)
+		_, hash, err := tree.Build(NewLoggerContextTodoForTesting(t), nil, kvps)
+		rootHashes[j] = hash
+		require.NoError(t, err)
+	}
+
+	verifier := NewMerkleProofVerifier(cfg)
+
+	startSeqno := Seqno(20)
+	endSeqno := Seqno(39)
+
+	// real proof verifies successfully
+	eProof, err := tree.GetExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, endSeqno)
+	require.NoError(t, err)
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), &eProof, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.NoError(t, err)
+
+	// nil proof
+	err = verifier.VerifyExtensionProof(NewLoggerContextTodoForTesting(t), nil, startSeqno, rootHashes[startSeqno], endSeqno, rootHashes[endSeqno])
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "nil proof")
```

Comment:
Is is possible to check the exact type of the error? I find this less fragile than checking for error strings.

Reply:
they are all ProofVerificationFailedErrors. I am checking the string to be more granular here. 
I think it isn't worth it to make sub error types, but I can add them if you want.

Reply:
Ok.  If they are all the same error then not worth it. Can you tell the two hash mismatches apart?

Reply:
Ok.  If they are all the same error then not worth it. Can you tell the two hash mismatches apart?

Reply:
No, it comes from the same spot in the code (even though the proof has been altered in two different ways, the final result is still a hash mismatch at the last check). Will try to make this more clear in the test.